File: cpp/src/node.cpp
#include "core.h"
#include "llm_layer.h"
#include <iostream>
#include <sstream>
#include <vector>
#include <string>
#include <thread>
#include <chrono>
#include <mutex>

using namespace llm_mina_chain;

std::mutex blockchain_mutex;
bool mining_active = true;

void mining_thread(Blockchain* blockchain) {
    while (mining_active) {
        std::this_thread::sleep_for(std::chrono::seconds(5));
        
        std::lock_guard<std::mutex> lock(blockchain_mutex);
        if (!blockchain->transaction_pool.empty()) {
            std::vector<Transaction> txs = blockchain->transaction_pool;
            auto block = blockchain->create_block(txs);
            if (block) {
                std::cout << "⛏️  Mined block #" << block->height 
                          << " with " << block->transactions.size() << " transactions" << std::endl;
            } else {
                std::cout << "❌ Failed to create block - invalid transactions" << std::endl;
            }
        }
    }
}

void print_help() {
    std::cout << "📖 Available Commands:" << std::endl;
    std::cout << "   help              - Show this help" << std::endl;
    std::cout << "   state             - Show current blockchain state" << std::endl;
    std::cout << "   block [height]    - Show specific block or latest" << std::endl;
    std::cout << "   chain             - Show entire blockchain" << std::endl;
    std::cout << "   transfer <s> <r> <a>  - Create transfer transaction" << std::endl;
    std::cout << "   gasless <s> <r> <a>   - Create gasless transaction" << std::endl;
    std::cout << "   llm <text>        - Parse natural language to transaction" << std::endl;
    std::cout << "   mine              - Mine next block" << std::endl;
    std::cout << "   pool              - Show transaction pool" << std::endl;
    std::cout << "   gas [price]       - Set or get gas price" << std::endl;
    std::cout << "   exit              - Exit the node" << std::endl;
}

void print_state(const Blockchain& bc) {
    std::cout << "   Balances:" << std::endl;
    for (const auto& [addr, balance] : bc.state.balances) {
        std::cout << "     " << addr << ": " << balance << std::endl;
    }
    std::cout << "   Nonces:" << std::endl;
    for (const auto& [addr, nonce] : bc.state.nonces) {
        std::cout << "     " << addr << ": " << nonce << std::endl;
    }
    std::cout << "   Gas Price: " << bc.gas_price << std::endl;
}

void print_block(const Block& block) {
    std::cout << "📦 Block #" << block.height << std::endl;
    std::cout << "   Hash: " << block.block_hash.substr(0, 64) << std::endl;
    std::cout << "   Previous: " << block.previous_hash.substr(0, 64) << std::endl;
    std::cout << "   State Hash: " << block.state_hash.substr(0, 64) << std::endl;
    std::cout << "   Proof: " << (block.proof ? block.proof.value() : "none") << std::endl;
    std::cout << "   Transactions: " << block.transactions.size() << std::endl;
    for (const auto& tx : block.transactions) {
        std::cout << "     " << tx.sender << " -> " << tx.receiver 
                  << " (" << tx.amount << ") [" << tx.tx_id.substr(0, 8) << "]" << std::endl;
    }
}

void print_chain(const Blockchain& bc) {
    std::cout << "🔗 Blockchain (" << bc.chain.size() << " blocks)" << std::endl;
    for (const auto& block : bc.chain) {
        std::cout << "   Block #" << block.height << ": " << block.transactions.size() 
                  << " transactions, hash: " << block.block_hash.substr(0, 16) << std::endl;
    }
}

void handle_command(const std::string& input, Blockchain* blockchain, const LLMTransactionParser& parser) {
    std::istringstream iss(input);
    std::vector<std::string> parts((std::istream_iterator<std::string>(iss)),
                                   std::istream_iterator<std::string>());
    
    if (parts.empty()) return;
    
    std::string cmd = parts[0];
    
    if (cmd == "help") {
        print_help();
    } else if (cmd == "state") {
        std::lock_guard<std::mutex> lock(blockchain_mutex);
        print_state(*blockchain);
    } else if (cmd == "block") {
        std::lock_guard<std::mutex> lock(blockchain_mutex);
        if (parts.size() >= 2) {
            try {
                uint64_t height = std::stoull(parts[1]);
                auto block = blockchain->get_block(height);
                if (block) {
                    print_block(*block);
                } else {
                    std::cout << "❌ Block #" << height << " not found" << std::endl;
                }
            } catch (...) {
                std::cout << "❌ Invalid block height" << std::endl;
            }
        } else {
            print_block(blockchain->get_latest_block());
        }
    } else if (cmd == "chain") {
        std::lock_guard<std::mutex> lock(blockchain_mutex);
        print_chain(*blockchain);
    } else if (cmd == "transfer") {
        if (parts.size() >= 4) {
            std::string sender = parts[1];
            std::string receiver = parts[2];
            uint64_t amount = std::stoull(parts[3]);
            
            std::lock_guard<std::mutex> lock(blockchain_mutex);
            uint64_t nonce = blockchain->state.get_nonce(sender);
            
            Transaction tx(sender, receiver, amount, nonce, 21000, 1);
            
            if (blockchain->add_transaction(tx)) {
                std::cout << "✅ Transaction added to pool: " << tx.tx_id << std::endl;
                std::cout << "   " << sender << " -> " << receiver << " (" << amount << ")" << std::endl;
            } else {
                std::cout << "❌ Transaction validation failed" << std::endl;
            }
        } else {
            std::cout << "Usage: transfer <sender> <receiver> <amount>" << std::endl;
        }
    } else if (cmd == "gasless") {
        if (parts.size() >= 4) {
            std::string sender = parts[1];
            std::string receiver = parts[2];
            uint64_t amount = std::stoull(parts[3]);
            
            std::lock_guard<std::mutex> lock(blockchain_mutex);
            uint64_t nonce = blockchain->state.get_nonce(sender);
            
            Transaction tx(sender, receiver, amount, nonce, std::nullopt, std::nullopt);
            
            if (blockchain->add_transaction(tx)) {
                std::cout << "✅ Gasless transaction added to pool: " << tx.tx_id << std::endl;
                std::cout << "   " << sender << " -> " << receiver << " (" << amount << ")" << std::endl;
            } else {
                std::cout << "❌ Transaction validation failed" << std::endl;
            }
        } else {
            std::cout << "Usage: gasless <sender> <receiver> <amount>" << std::endl;
        }
    } else if (cmd == "llm") {
        if (parts.size() >= 2) {
            std::string text;
            for (size_t i = 1; i < parts.size(); ++i) {
                if (i > 1) text += " ";
                text += parts[i];
            }
            
            std::lock_guard<std::mutex> lock(blockchain_mutex);
            ParsedTransaction parsed = parser.parse(text, "alice");
            
            std::cout << "🤖 Parsed transaction:" << std::endl;
            std::cout << "   Confidence: " << parsed.confidence << std::endl;
            std::cout << "   Explanation: " << parsed.explanation << std::endl;
            std::cout << "   Sender: " << (parsed.sender ? parsed.sender.value() : "none") << std::endl;
            std::cout << "   Receiver: " << parsed.receiver << std::endl;
            std::cout << "   Amount: " << parsed.amount << std::endl;
            std::cout << "   Gas: " << (parsed.gas_limit ? std::to_string(*parsed.gas_limit) : "none") << std::endl;
            
            if (parsed.confidence > 0.5) {
                uint64_t nonce = blockchain->state.get_nonce(parsed.sender.value_or("alice"));
                Transaction tx(parsed.sender.value_or("alice"), parsed.receiver, 
                             parsed.amount, nonce, parsed.gas_limit, parsed.gas_price);
                
                if (blockchain->add_transaction(tx)) {
                    std::cout << "✅ Transaction added to pool: " << tx.tx_id << std::endl;
                } else {
                    std::cout << "❌ Transaction validation failed" << std::endl;
                }
            }
        } else {
            std::cout << "Usage: llm <natural language command>" << std::endl;
            std::cout << "Example: llm transfer 100 from alice to bob" << std::endl;
            std::cout << "Example: llm send 50 to bob gasless" << std::endl;
        }
    } else if (cmd == "mine") {
        std::lock_guard<std::mutex> lock(blockchain_mutex);
        std::vector<Transaction> txs = blockchain->transaction_pool;
        
        if (txs.empty()) {
            std::cout << "⚠️  No transactions in pool" << std::endl;
            return;
        }
        
        auto block = blockchain->create_block(txs);
        if (block) {
            std::cout << "⛏️  Mined block #" << block->height 
                      << " with " << block->transactions.size() << " transactions" << std::endl;
            print_block(*block);
        } else {
            std::cout << "❌ Failed to create block - invalid transactions" << std::endl;
        }
    } else if (cmd == "pool") {
        std::lock_guard<std::mutex> lock(blockchain_mutex);
        std::cout << "📦 Transaction Pool (" << blockchain->transaction_pool.size() << " transactions)" << std::endl;
        for (const auto& tx : blockchain->transaction_pool) {
            std::cout << "   " << tx.sender << " -> " << tx.receiver 
                      << " (" << tx.amount << ") [" << tx.tx_id.substr(0, 8) << "]" << std::endl;
        }
    } else if (cmd == "gas") {
        if (parts.size() >= 2) {
            try {
                uint64_t price = std::stoull(parts[1]);
                std::lock_guard<std::mutex> lock(blockchain_mutex);
                blockchain->set_gas_price(price);
                std::cout << "⛽ Gas price set to " << price << std::endl;
            } catch (...) {
                std::cout << "❌ Invalid gas price" << std::endl;
            }
        } else {
            std::lock_guard<std::mutex> lock(blockchain_mutex);
            std::cout << "⛽ Current gas price: " << blockchain->get_gas_price() << std::endl;
        }
    } else if (cmd == "exit" || cmd == "quit") {
        std::cout << "👋 Goodbye!" << std::endl;
        mining_active = false;
        std::exit(0);
    } else {
        std::cout << "❓ Unknown command. Type 'help' for available commands." << std::endl;
    }
}

int main() {
    std::cout << "🔗 LLM-Mina-Chain Node v0.1.0" << std::endl;
    std::cout << "================================" << std::endl << std::endl;
    
    // Initialize blockchain
    Blockchain blockchain;
    LLMTransactionParser parser;
    
    std::cout << "✅ Blockchain initialized" << std::endl;
    std::cout << "📊 Current state:" << std::endl;
    print_state(blockchain);
    
    // Start mining thread
    std::thread miner(mining_thread, &blockchain);
    miner.detach();
    
    // Main loop
    std::string input;
    while (std::getline(std::cin, input)) {
        input.erase(0, input.find_first_not_of(" \t\n\r"));
        input.erase(input.find_last_not_of(" \t\n\r") + 1);
        
        if (input.empty()) continue;
        
        handle_command(input, &blockchain, parser);
    }
    
    mining_active = false;
    return 0;
}

File: rust/src/node.rs
//! Blockchain node for LLM-Mina-Chain

use llm_mina_chain::{Blockchain, Transaction, LLMTransactionParser};
use std::io::{self, BufRead};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

fn main() {
    println!("🔗 LLM-Mina-Chain Node v0.1.0");
    println!("================================\n");
    
    // Initialize blockchain
    let blockchain = Arc::new(Mutex::new(Blockchain::new()));
    let parser = LLMTransactionParser::new();
    
    println!("✅ Blockchain initialized");
    println!("📊 Current state:");
    print_state(&blockchain.lock().unwrap());
    
    // Start mining thread
    let blockchain_clone = blockchain.clone();
    thread::spawn(move || {
        loop {
            thread::sleep(Duration::from_secs(5));
            let mut bc = blockchain_clone.lock().unwrap();
            
            if !bc.transaction_pool.is_empty() {
                let txs = bc.transaction_pool.clone();
                match bc.create_block(txs) {
                    Some(block) => {
                        println!("⛏️  Mined block #{} with {} transactions", block.height, block.transactions.len());
                    }
                    None => {
                        println!("❌ Failed to create block - invalid transactions");
                    }
                }
            }
        }
    });
    
    // Main loop
    let stdin = io::stdin();
    for line in stdin.lock().lines() {
        let input = line.unwrap();
        let input = input.trim();
        
        if input.is_empty() {
            continue;
        }
        
        handle_command(input, &blockchain, &parser);
    }
}

fn handle_command(input: &str, blockchain: &Arc<Mutex<Blockchain>>, parser: &LLMTransactionParser) {
    let parts: Vec<&str> = input.split_whitespace().collect();
    
    match parts.get(0).map(|s| *s) {
        Some("help") => print_help(),
        Some("state") => {
            let bc = blockchain.lock().unwrap();
            print_state(&bc);
        }
        Some("block") => {
            if let Some(height_str) = parts.get(1) {
                if let Ok(height) = height_str.parse::<u64>() {
                    let bc = blockchain.lock().unwrap();
                    if let Some(block) = bc.get_block(height) {
                        print_block(block);
                    } else {
                        println!("❌ Block #{} not found", height);
                    }
                }
            } else {
                let bc = blockchain.lock().unwrap();
                print_block(bc.get_latest_block());
            }
        }
        Some("chain") => {
            let bc = blockchain.lock().unwrap();
            print_chain(&bc);
        }
        Some("transfer") => {
            if parts.len() >= 4 {
                let sender = parts[1];
                let receiver = parts[2];
                let amount: u64 = parts[3].parse().unwrap_or(0);
                
                let mut bc = blockchain.lock().unwrap();
                let nonce = bc.state.get_nonce(sender);
                
                let tx = Transaction::new(
                    sender.to_string(),
                    receiver.to_string(),
                    amount,
                    nonce,
                    Some(21000),
                    Some(1),
                );
                
                if bc.add_transaction(tx.clone()) {
                    println!("✅ Transaction added to pool: {}", tx.tx_id);
                    println!("   {} -> {} ({})", sender, receiver, amount);
                } else {
                    println!("❌ Transaction validation failed");
                }
            } else {
                println!("Usage: transfer <sender> <receiver> <amount>");
            }
        }
        Some("gasless") => {
            if parts.len() >= 4 {
                let sender = parts[1];
                let receiver = parts[2];
                let amount: u64 = parts[3].parse().unwrap_or(0);
                
                let mut bc = blockchain.lock().unwrap();
                let nonce = bc.state.get_nonce(sender);
                
                let tx = Transaction::new(
                    sender.to_string(),
                    receiver.to_string(),
                    amount,
                    nonce,
                    None,  // gasless
                    None,
                );
                
                if bc.add_transaction(tx.clone()) {
                    println!("✅ Gasless transaction added to pool: {}", tx.tx_id);
                    println!("   {} -> {} ({})", sender, receiver, amount);
                } else {
                    println!("❌ Transaction validation failed");
                }
            } else {
                println!("Usage: gasless <sender> <receiver> <amount>");
            }
        }
        Some("llm") => {
            if parts.len() >= 2 {
                let text = parts[1..].join(" ");
                let bc = blockchain.lock().unwrap();
                
                let parsed = parser.parse(&text, Some("alice"));
                println!("🤖 Parsed transaction:");
                println!("   Confidence: {:.1}", parsed.confidence);
                println!("   Explanation: {}", parsed.explanation);
                println!("   Sender: {:?}", parsed.sender);
                println!("   Receiver: {}", parsed.receiver);
                println!("   Amount: {}", parsed.amount);
                println!("   Gas: {:?}", parsed.gas_limit);
                
                if parsed.confidence > 0.5 {
                    let nonce = bc.state.get_nonce(parsed.sender.as_deref().unwrap_or("alice"));
                    let tx = Transaction::new(
                        parsed.sender.unwrap_or_else(|| "alice".to_string()),
                        parsed.receiver,
                        parsed.amount,
                        nonce,
                        parsed.gas_limit,
                        parsed.gas_price,
                    );
                    
                    drop(bc);
                    let mut bc = blockchain.lock().unwrap();
                    if bc.add_transaction(tx.clone()) {
                        println!("✅ Transaction added to pool: {}", tx.tx_id);
                    } else {
                        println!("❌ Transaction validation failed");
                    }
                }
            } else {
                println!("Usage: llm <natural language command>");
                println!("Example: llm transfer 100 from alice to bob");
                println!("Example: llm send 50 to bob gasless");
            }
        }
        Some("mine") => {
            let mut bc = blockchain.lock().unwrap();
            let txs = bc.transaction_pool.clone();
            
            if txs.is_empty() {
                println!("⚠️  No transactions in pool");
                return;
            }
            
            match bc.create_block(txs) {
                Some(block) => {
                    println!("⛏️  Mined block #{} with {} transactions", block.height, block.transactions.len());
                    print_block(&block);
                }
                None => {
                    println!("❌ Failed to create block - invalid transactions");
                }
            }
        }
        Some("pool") => {
            let bc = blockchain.lock().unwrap();
            println!("📦 Transaction Pool ({} transactions)", bc.transaction_pool.len());
            for tx in &bc.transaction_pool {
                println!("   {} -> {} ({}) [{}]", tx.sender, tx.receiver, tx.amount, &tx.tx_id[..8]);
            }
        }
        Some("gas") => {
            if let Some(price_str) = parts.get(1) {
                if let Ok(price) = price_str.parse::<u64>() {
                    let mut bc = blockchain.lock().unwrap();
                    bc.set_gas_price(price);
                    println!("⛽ Gas price set to {}", price);
                }
            } else {
                let bc = blockchain.lock().unwrap();
                println!("⛽ Current gas price: {}", bc.get_gas_price());
            }
        }
        Some("exit") | Some("quit") => {
            println!("👋 Goodbye!");
            std::process::exit(0);
        }
        _ => {
            println!("❓ Unknown command. Type 'help' for available commands.");
        }
    }
}

fn print_help() {
    println!("📖 Available Commands:");
    println!("   help              - Show this help");
    println!("   state             - Show current blockchain state");
    println!("   block [height]    - Show specific block or latest");
    println!("   chain             - Show entire blockchain");
    println!("   transfer <s> <r> <a>  - Create transfer transaction");
    println!("   gasless <s> <r> <a>   - Create gasless transaction");
    println!("   llm <text>        - Parse natural language to transaction");
    println!("   mine              - Mine next block");
    println!("   pool              - Show transaction pool");
    println!("   gas [price]       - Set or get gas price");
    println!("   exit              - Exit the node");
}

fn print_state(bc: &Blockchain) {
    println!("   Balances:");
    for (addr, balance) in &bc.state.balances {
        println!("     {}: {}", addr, balance);
    }
    println!("   Nonces:");
    for (addr, nonce) in &bc.state.nonces {
        println!("     {}: {}", addr, nonce);
    }
    println!("   Gas Price: {}", bc.gas_price);
}

fn print_block(block: &llm_mina_chain::Block) {
    println!("📦 Block #{}", block.height);
    println!("   Hash: {}", &block.block_hash[..64.min(block.block_hash.len())]);
    println!("   Previous: {}", &block.previous_hash[..64.min(block.previous_hash.len())]);
    println!("   State Hash: {}", &block.state_hash[..64.min(block.state_hash.len())]);
    println!("   Proof: {:?}", block.proof);
    println!("   Transactions: {}", block.transactions.len());
    for tx in &block.transactions {
        println!("     {} -> {} ({}) [{}]", tx.sender, tx.receiver, tx.amount, &tx.tx_id[..8]);
    }
}

fn print_chain(bc: &Blockchain) {
    println!("🔗 Blockchain ({} blocks)", bc.chain.len());
    for block in &bc.chain {
        println!("   Block #{}: {} transactions, hash: {}", 
            block.height, 
            block.transactions.len(),
            &block.block_hash[..16.min(block.block_hash.len())]
        );
    }
}
